### PR TITLE
Add custom errors to search

### PR DESCRIPTION
**Pull request by GiancarloFusiello**
relates to issue [108](https://github.com/libero/libero/issues/108)

### CODE REVIEW

File: tests/test_errors.py
Comment (GiancarloFusiello) [inline tests/test_errors.py]: tests are currently failing for error code 412. I've opened an issue [here](https://github.com/pallets/flask/issues/3093)

Comment (GiancarloFusiello) [reply tests/test_errors.py]: This appears to be by design. @thewilkybarkid should I just comment this out or remove 412 from the test?

Comment (thewilkybarkid) [reply tests/test_errors.py]: Technically we won't need this code for this service (since it's read only), but we will want it for future writeable apps.

Comment (GiancarloFusiello) [reply tests/test_errors.py]: So should I comment it out or remove it from the test?

File: schemas/error.rng
Comment (giorgiosironi) [inline schemas/error.rng]: wonder where we can put a link to the original file, since this is probably lifted from `schemas` but it's not obvious

File: search/__init__.py
Comment (giorgiosironi) [inline search/__init__.py]: this is hardcoding the name of one environment, so it may make difficult to reuse the application across custom environments (but it's just `development`); I think we are after a setting named something like `APP_ERROR_BLUEPRINT` or `APP_ERROR_ROUTES` that can be turned on or off. This can be turned on in `development` and be off by default.

How do we turn on a setting in development? By putting a `env:` line in `docker-compose-dev.yml` that passes in a setting.

File: search/errors.py
Comment (giorgiosironi) [inline search/errors.py]: that looks like a awkward API that we are forced to call for the creation of the document; @gnott do you have suggestions for improving the ease with which a small XML document can be created with `lxml`?

(I'd still avoid harcoding a XML string if it has to have dynamic values in it like these fields)

Comment (giorgiosironi) [reply search/errors.py]: e.g. the [`lxml.builder` module](https://lxml.de/api/lxml.builder.ElementMaker-class.html)? Sample for HTML, but it's similar:
```
page = (
    E.html(
        E.head(
            E.title("This is a sample document")
        ),
        E.body(
            E.h1("Hello!", CLASS("title")),
            E.p("This is a paragraph with ", B("bold"), " text in it!"),
            E.p("This is another paragraph, with a ",
                A("link", href="http://www.python.org"), "."),
            E.p("Here are some reserved characters: <spam&egg>."),
            ET.XML("<p>And finally, here is an embedded XHTML fragment.</p>"),
        )
    )
)
```

File: search/__init__.py
Comment (GiancarloFusiello) [reply search/__init__.py]: There is a config setting entitled: 'TESTING' which is automatically set to True when the tests are run. I'll update the code to use this as the `/error` endpoint was only implemented as method to test error responses.

File: schemas/error.rng
Comment (GiancarloFusiello) [reply schemas/error.rng]: I can try to pull the `libero/schemas` repo and see if I can read it directly. What do you think?

File: search/errors.py
Comment (GiancarloFusiello) [reply search/errors.py]: @giorgiosironi Are you referring to the namespace on line 11?

File: tests/test_errors.py
Comment (GiancarloFusiello) [reply tests/test_errors.py]: I've commented this out for now and left a small explanation, referencing the flask issue I raised.

File: schemas/error.rng
Comment (GiancarloFusiello) [reply schemas/error.rng]: I'm going to create an issue to do this separately after this PR is merged.

File: search/__init__.py
Comment (GiancarloFusiello) [reply search/__init__.py]: I have moved the registration of the error blueprint to the pytest fixture where the test client is created, removing the need to check a config variable.

File: search/errors.py
Comment (giorgiosironi) [reply search/errors.py]: To the fact that each element needs to be created as a `SubElement` object and then its `text` field set separately, can be annoying to set up a (longer) document this way?

File: search/xml.py
Comment (thewilkybarkid) [inline search/xml.py]: `_Validator` as a simpler return type?

**Comment by GiancarloFusiello** [comment]
I'm going to park this until [113](https://github.com/libero/libero/issues/113) has been completed as there are overlapping sections that will result in refactoring this PR.

File: search/errors.py
Comment (GiancarloFusiello) [reply search/errors.py]: updated.

File: search/xml.py
Comment (GiancarloFusiello) [reply search/xml.py]: I have since removed this file as these functions are no longer needed.

**Comment by GiancarloFusiello** [comment]
@giorgiosironi @thewilkybarkid No longer blocked and ready for review.

File: search/errors.py
Comment (giorgiosironi) [inline search/errors.py:30]: I remember this case before, I can't figure out which is the unused variable? Only `status_code` is assigned in this function, and it is used in `abort()`

Comment (GiancarloFusiello) [reply search/errors.py:30]: It's the function name. In this case `def error()` is the unused variable name. Pylint thinks a function declaration within another function that is then not called within the outer function is unused. But, Flask does map the inner function after being registered which will then be called when the route is accessed/requested.

File: tests/test_errors.py
Comment (giorgiosironi) [inline tests/test_errors.py]: I consider conditionals in test a smell since the increase the complexity of the test method. `test_custom_error_handler` already has a 404 case and it's testing its full response, so I'd remove the `None` case from the `parametrize` cases and transform it into a focused test that doesn't go through the XML version but only checks this "default" aspect:
```
def test_default_error_code(self):
    response = client.get('/error')
    assert response.status_code == 404
```

File: search/errors.py
Comment (giorgiosironi) [inline search/errors.py:19]: looks clean now

File: tests/test_errors.py
Comment (GiancarloFusiello) [reply tests/test_errors.py]: updated.

File: search/__init__.py
Comment (thewilkybarkid) [inline search/__init__.py]: Is it not possible to say any 4xx/5xx code?

Assume any other exception ends up as a 500 and is still handled?

Comment (GiancarloFusiello) [reply search/__init__.py]: I don't think so. Here the error handler is registered for each error code. I agree that it would make sense to have something more encompassing but I don't think this is possible using nice syntax. An alternative is that I create a list of integers from 400 - 599 and register the custom handler for each integer. I'll have a closer look at the Flask code and see if I can find a work around.

Comment (GiancarloFusiello) [reply search/__init__.py]: After reading the library code directly, I was able to do this. Pushing an update.

File: tests/test_errors.py
Comment (GiancarloFusiello) [inline tests/test_errors.py:24]: @thewilkybarkid @giorgiosironi Flask currently has a dependency of `werkzeug >=0.14`. I can confirm this is fixed in version `0.15.dev0` which hopefully indicates that a release of `0.15` is imminent. Is it worth using `0.15.dev0` now or should we wait for an official release? I'm hesitant to use this in production but since this is not in production, what do you think?

Comment (giorgiosironi) [reply tests/test_errors.py:24]: A comment capturing this knowledge plus postponing the upgrade is fine for me, since we are not using this code at the moment fixing it is technically unnecessary.

**Review by giorgiosironi — APPROVED** [review]

**Pull request merged**
